File: Combination Sum/Combination_Sum.py
# 76ms 81.65%
class Solution:
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        # 把reslist写在前面会大幅度提高处理速度
        res_list, len_nums = [], len(candidates)
        candidates.sort()
        
        def mysum(target, temp_list, start_index = 0):
            for index in range(start_index, len_nums):
                temp = target - candidates[index]
                if temp is 0:
                    res_list.append(temp_list + [candidates[index]])
                elif temp < 0:
                    break
                else:
                    # 这里为了去重，增加了一个start_index
                    mysum(temp, temp_list + [candidates[index]], index)
        
        mysum(target, [])
        return res_list

File: Combination Sum/Combination_Sum.cpp
// 典型的回溯法
// Runtime: 8 ms, faster than 98.32% of C++ online submissions for Combination Sum.
// Memory Usage: 9.4 MB, less than 80.28% of C++ online submissions for Combination Sum.

class Solution 
{
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) 
    {
        sort(candidates.begin(), candidates.end());
        if (candidates.size() == 0 || candidates[0] > target)
            return vector<vector<int>>();
        
        vector<vector<int>> res;
        vector<int> temp;
        
        helper(candidates, res, temp, target, 0, 0);
        return res;
    }
private:
    void helper(vector<int>& candidates, vector<vector<int>>& res, vector<int>& temp, int target, int curSum, int start)
    {
        if (curSum == target)
        {
            res.push_back(temp);
            return ;
        }
        
        for (int i = start; i < candidates.size(); ++i)
        {
            if (curSum + candidates[i] > target)
                break;
            
            curSum += candidates[i];
            temp.push_back(candidates[i]);
            
            helper(candidates, res, temp, target, curSum, i);
            
            curSum -= candidates[i];
            temp.pop_back();
        }
    }
};